atoma-proxy/src/server/middleware.rs

```diff
@@ -2211,7 +2327,136 @@ pub mod utils {
             user_id,
-            selected_stack_small_id,
+            selected_stack_small_id: Some(selected_stack_small_id),
+            fiat_estimated_amount: None,
+            price_per_million: None,
+            endpoint: endpoint.to_string(),
+            model_name: request_model.to_string(),
+        });
+
+        // update headers
+        let req = Request::from_parts(req_parts.clone(), Body::from(body_json.to_string()));
+        Ok(req)
+    }
+
+    /// Validates and prepares a request for processing by a fiat request and node.
+    ///
+    /// This function performs several key operations to prepare a request for forwarding:
+    /// 1. Obtain node address and signature
+    /// 2. Sets up required headers for node communication
+    /// 3. Configures request metadata for tracking and routing
+    ///
+    /// # Arguments
+    ///
+    /// * `state` - Server state containing shared resources and connections
+    /// * `body_json` - The parsed JSON body of the request
+    /// * `req_parts` - Mutable reference to request parts for header modification
+    /// * `selected_node_id` - ID of the node selected to process this request
+    /// * `fiat_estimated_amount` - The amount of fiat that was locked
+    /// * `total_compute_units` - Total compute units required for this request
+    /// * `user_id` - ID of the user making the request
+    /// * `endpoint` - API endpoint path being accessed
+    ///
+    /// # Returns
+    ///
+    /// Returns a `Result<Request<Body>>` containing the fully prepared request if successful.
+    ///
+    /// # Errors
+    ///
+    /// Returns `AtomaProxyError` in the following cases:
+    /// * `InternalError` - Failed to:
+    ///   - Convert values to header format
+    ///   - Set up required headers
+    /// * `InvalidBody` - Request body missing required "model" field
+    ///
+    /// # Headers Set
+    ///
+    /// The following headers are set on the request:
+    /// * `X-Signature` - Authentication signature for the node
+    /// * `Content-Length` - Updated body length
+    /// * `X-Tx-Digest` - (Optional) Transaction digest for new stacks
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// let prepared_request = create_fiat_request(
+    ///     &state,
+    ///     &body_json,
+    ///     &mut req_parts,
+    ///     node_id,
+    ///     fiat_estimated_amount,
+    ///     compute_units,
+    ///     user_id,
+    ///     "/v1/chat/completions"
+    /// ).await?;
+    /// ```
+    ///
+    /// # Request Metadata
+    ///
+    /// The function also sets up `RequestMetadataExtension` with:
+    /// * Node address and ID
+    /// * Compute units allocation
+    /// * Stack ID
+    /// * Endpoint path
+    /// * Model name
+    #[instrument(level = "info", skip_all, fields(
+        %endpoint,
+        %fiat_estimated_amount,
+        %user_id
+    ), err)]
+    #[allow(clippy::too_many_arguments)]
+    pub async fn create_fiat_request(
```

Comment:
Do we want to avoid replay attacks when these come not from the proxy ?

Reply:
Yes, don't know how yet. But how would an attacker get hands on such request? It's not public.